### PR TITLE
[SPARK-9142][SQL] Removing unnecessary self types in Catalyst.

### PR DESCRIPTION
**Pull request by rxin**
Just a small change to add Product type to the base expression/plan abstract classes, based on suggestions on #7434 and offline discussions.

### CODE REVIEW

**Comment by marmbrus** [comment]
LGTM

**Comment by marmbrus** [comment]
Merging to master.  Good riddance to `self: Product =>`!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -25,8 +25,7 @@ import org.apache.spark.sql.catalyst.plans.QueryPlan
 import org.apache.spark.sql.catalyst.trees.TreeNode
 
 
-abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
-  self: Product =>
+abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging with Product{
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: why not making `QueryPlan` extends `Product`?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: or why not just make `TreeNode` extends `Product`?

**Pull request closed without merging**